File: tools/merge_sprites.py
#!/usr/bin/env python3
"""
怪物精灵图表合并脚本
将 PNG 序列合并为 sprite sheets
"""

import os
import sys
from PIL import Image

def get_image_size(folder):
    """获取文件夹中第一个图片的尺寸"""
    for file in os.listdir(folder):
        if file.endswith('.png'):
            img = Image.open(os.path.join(folder, file))
            return img.size
    return (64, 64)

def merge_animation_frames(folder, output_prefix):
    """合并动画帧为水平 sprite sheet"""
    if not os.path.exists(folder):
        print(f"跳过：{folder} 不存在")
        return []
    
    # 获取所有 PNG 文件并排序
    files = sorted([f for f in os.listdir(folder) if f.endswith('.png')])
    
    if not files:
        print(f"跳过：{folder} 没有 PNG 文件")
        return []
    
    # 获取单帧尺寸
    first_img = Image.open(os.path.join(folder, files[0]))
    frame_width, frame_height = first_img.size
    
    # 创建合并的图片（水平排列）
    total_width = frame_width * len(files)
    merged_img = Image.new('RGBA', (total_width, frame_height), (0, 0, 0, 0))
    
    # 粘贴所有帧
    for i, file in enumerate(files):
        img = Image.open(os.path.join(folder, file))
        merged_img.paste(img, (i * frame_width, 0))
    
    # 保存
    output_path = f"{output_prefix}.png"
    merged_img.save(output_path)
    
    print(f"✓ {output_path} ({len(files)}帧，{total_width}x{frame_height})")
    
    return {
        'path': output_path,
        'frames': len(files),
        'frame_width': frame_width,
        'frame_height': frame_height,
        'total_width': total_width
    }

def process_monster(monster_name, base_path='assets/monsters'):
    """处理单个怪物的所有动画"""
    print(f"\n处理怪物：{monster_name}")
    
    monster_path = os.path.join(base_path, 'PNG', monster_name)
    output_path = os.path.join(base_path, monster_name)
    
    os.makedirs(output_path, exist_ok=True)
    
    animations = {}
    
    # 动画类型映射
    animation_types = {
        'Idle': 'idle',
        'Walk': 'walk',
        'Attack': 'attack',
        'Hurt': 'hurt',
        'Death': 'death',
        'Fire_Attack': 'fire_attack',
        'Magic_Attack': 'magic_attack',
        'Flight': 'flight',
        'Stone': 'stone'
    }
    
    for anim_name, anim_key in animation_types.items():
        anim_folder = os.path.join(monster_path, anim_name)
        if os.path.exists(anim_folder):
            result = merge_animation_frames(
                anim_folder,
                os.path.join(output_path, anim_key)
            )
            if result:
                animations[anim_key] = result
    
    return animations

def main():
    base_path = 'assets/monsters'
    png_path = os.path.join(base_path, 'PNG')
    
    if not os.path.exists(png_path):
        print(f"错误：{png_path} 不存在")
        sys.exit(1)
    
    # 获取所有怪物文件夹
    monsters = [d for d in os.listdir(png_path) if os.path.isdir(os.path.join(png_path, d))]
    
    print(f"发现 {len(monsters)} 个怪物：{', '.join(monsters)}")
    
    all_animations = {}
    
    for monster in monsters:
        animations = process_monster(monster, base_path)
        all_animations[monster] = animations
    
    # 生成配置 JSON
    import json
    config = {
        'monsters': {}
    }
    
    for monster, animations in all_animations.items():
        config['monsters'][monster] = {}
        for anim_name, anim_data in animations.items():
            config['monsters'][monster][anim_name] = {
                'texture': anim_data['path'],
                'frames': anim_data['frames'],
                'frameWidth': anim_data['frame_width'],
                'frameHeight': anim_data['frame_height']
            }
    
    # 保存配置
    config_path = os.path.join(base_path, 'monster_animations.json')
    with open(config_path, 'w', encoding='utf-8') as f:
        json.dump(config, f, indent=2, ensure_ascii=False)
    
    print(f"\n✓ 配置已保存到：{config_path}")
    print(f"\n完成！共处理 {len(monsters)} 个怪物")

if __name__ == '__main__':
    main()

File: tools/merge_sprites.js
#!/usr/bin/env node
/**
 * 怪物精灵图表合并脚本
 * 将 PNG 序列合并为 sprite sheets
 * 支持两种结构:
 * 1. 扁平结构：Attack1.png, Attack2.png...
 * 2. 文件夹结构：Attack/Attack1.png, Attack/Attack2.png...
 */

const fs = require('fs');
const path = require('path');

// 使用 canvas 库
const { createCanvas, loadImage } = require('canvas');

function extractAnimationType(filename) {
    // 从文件名提取动画类型，如 "Attack1.png" -> "Attack"
    const match = filename.match(/^([A-Za-z_]+)\d+\.png$/);
    return match ? match[1] : null;
}

function groupFilesByAnimation(files) {
    const groups = {};
    
    for (const file of files) {
        const animType = extractAnimationType(file);
        if (animType) {
            if (!groups[animType]) {
                groups[animType] = [];
            }
            groups[animType].push(file);
        }
    }
    
    // 排序每个组
    for (const animType in groups) {
        groups[animType].sort((a, b) => {
            const numA = parseInt(a.match(/\d+/)[0]);
            const numB = parseInt(b.match(/\d+/)[0]);
            return numA - numB;
        });
    }
    
    return groups;
}

async function mergeAnimationFrames(files, folder, outputPrefix) {
    if (files.length === 0) {
        return null;
    }
    
    // 加载第一帧获取尺寸
    const firstImg = await loadImage(path.join(folder, files[0]));
    const frameWidth = firstImg.width;
    const frameHeight = firstImg.height;
    
    // 创建画布（水平排列所有帧）
    const totalWidth = frameWidth * files.length;
    const canvas = createCanvas(totalWidth, frameHeight);
    const ctx = canvas.getContext('2d');
    
    // 绘制所有帧
    for (let i = 0; i < files.length; i++) {
        const img = await loadImage(path.join(folder, files[i]));
        ctx.drawImage(img, i * frameWidth, 0);
    }
    
    // 保存
    const outputPath = `${outputPrefix}.png`;
    const buffer = canvas.toBuffer('image/png');
    fs.writeFileSync(outputPath, buffer);
    
    console.log(`✓ ${path.basename(outputPath)} (${files.length}帧，${totalWidth}x${frameHeight})`);
    
    return {
        path: outputPath,
        frames: files.length,
        frameWidth,
        frameHeight,
        totalWidth
    };
}

async function processMonster(monsterName, basePath = 'assets/monsters') {
    console.log(`\n处理怪物：${monsterName}`);
    
    const monsterPath = path.join(basePath, 'PNG', monsterName);
    const outputPath = path.join(basePath, monsterName);
    
    if (!fs.existsSync(outputPath)) {
        fs.mkdirSync(outputPath, { recursive: true });
    }
    
    // 获取所有 PNG 文件
    const files = fs.readdirSync(monsterPath).filter(f => f.endsWith('.png'));
    
    if (files.length === 0) {
        console.log(`  跳过：没有 PNG 文件`);
        return {};
    }
    
    // 按动画类型分组
    const groups = groupFilesByAnimation(files);
    
    console.log(`  发现 ${Object.keys(groups).length} 种动画：${Object.keys(groups).join(', ')}`);
    
    const animations = {};
    
    // 动画类型映射（标准化命名）
    const animationMap = {
        'Idle': 'idle',
        'Walk': 'walk',
        'Attack': 'attack',
        'Hurt': 'hurt',
        'Death': 'death',
        'Fire_Attack': 'fire_attack',
        'Magic_Attack': 'magic_attack',
        'Flight': 'flight',
        'Stone': 'stone'
    };
    
    for (const [animName, files] of Object.entries(groups)) {
        const animKey = animationMap[animName] || animName.toLowerCase();
        const result = await mergeAnimationFrames(
            files,
            monsterPath,
            path.join(outputPath, animKey)
        );
        if (result) {
            animations[animKey] = result;
        }
    }
    
    return animations;
}

async function main() {
    const basePath = 'assets/monsters';
    const pngPath = path.join(basePath, 'PNG');
    
    if (!fs.existsSync(pngPath)) {
        console.error(`错误：${pngPath} 不存在`);
        process.exit(1);
    }
    
    // 获取所有怪物文件夹
    const monsters = fs.readdirSync(pngPath)
        .filter(d => fs.statSync(path.join(pngPath, d)).isDirectory());
    
    console.log(`发现 ${monsters.length} 个怪物：${monsters.join(', ')}`);
    
    const allAnimations = {};
    
    for (const monster of monsters) {
        const animations = await processMonster(monster, basePath);
        allAnimations[monster] = animations;
    }
    
    // 生成配置 JSON
    const config = {
        monsters: {}
    };
    
    for (const [monster, animations] of Object.entries(allAnimations)) {
        config.monsters[monster] = {};
        for (const [animName, animData] of Object.entries(animations)) {
            config.monsters[monster][animName] = {
                texture: animData.path,
                frames: animData.frames,
                frameWidth: animData.frameWidth,
                frameHeight: animData.frameHeight
            };
        }
    }
    
    // 保存配置
    const configPath = path.join(basePath, 'monster_animations.json');
    fs.writeFileSync(configPath, JSON.stringify(config, null, 2), 'utf-8');
    
    console.log(`\n✓ 配置已保存到：${configPath}`);
    console.log(`\n完成！共处理 ${monsters.length} 个怪物`);
}

main().catch(console.error);
